File: projhub/lib/filetree.py
from __future__ import annotations
import json
from pathlib import Path

_SKIP_DIRS = {
    "node_modules", "dist", "build", ".next", "out", "target",
    "__pycache__", ".venv", "venv", ".tox", "coverage",
    ".turbo", ".parcel-cache", ".cache",
}

_BADGE_RULES = [
    {"files": [".git"], "badge": "git", "label": "Git"},
    {"files": ["package.json"], "badge": "node", "label": "Node", "detect": "_detect_node"},
    {"files": ["pyproject.toml", "requirements.txt", "setup.py"], "badge": "python", "label": "Python"},
    {"files": ["go.mod"], "badge": "go", "label": "Go"},
    {"files": ["Cargo.toml"], "badge": "rust", "label": "Rust"},
    {"files": ["pubspec.yaml"], "badge": "flutter", "label": "Flutter"},
    {"files": ["build.gradle", "pom.xml", "build.gradle.kts"], "badge": "java", "label": "Java"},
    {"files": ["composer.json"], "badge": "php", "label": "PHP"},
    {"files": ["Dockerfile", "docker-compose.yml", "docker-compose.yaml"], "badge": "docker", "label": "Docker"},
]


def _detect_node(dir_path: Path) -> str:
    try:
        pkg = json.loads((dir_path / "package.json").read_text(encoding="utf-8"))
        deps = {**pkg.get("dependencies", {}), **pkg.get("devDependencies", {})}
        if "react" in deps or "next" in deps:
            return "React"
        if "vue" in deps or "nuxt" in deps:
            return "Vue"
        if "svelte" in deps or "@sveltejs/kit" in deps:
            return "Svelte"
        if "express" in deps or "fastify" in deps or "hono" in deps:
            return "Server"
        if "react-native" in deps or "expo" in deps:
            return "React Native"
        if "electron" in deps:
            return "Electron"
    except Exception:
        pass
    return "Node"


def _detect_badges(dir_path: Path) -> list[dict]:
    try:
        entries = {e.name.lower() for e in dir_path.iterdir()}
    except Exception:
        entries = set()

    badges = []
    for rule in _BADGE_RULES:
        hit = any(f.lower() in entries for f in rule["files"])
        if hit:
            label = _detect_node(dir_path) if rule.get("detect") == "_detect_node" else rule["label"]
            badges.append({"badge": rule["badge"], "label": label})

    # Terraform .tf files
    if not any(b["badge"] == "terraform" for b in badges):
        if any(e.endswith(".tf") for e in entries):
            badges.append({"badge": "terraform", "label": "Terraform"})

    # Xcode
    if any(e.endswith(".xcodeproj") or e.endswith(".xcworkspace") for e in entries):
        badges.append({"badge": "ios", "label": "iOS"})

    return badges


def scan_dir(abs_path: Path, depth: int = 0, max_depth: int = 1, skip_hidden: bool = False) -> list[dict]:
    if not abs_path.exists():
        return []

    try:
        entries = list(abs_path.iterdir())
    except Exception:
        return []

    result = []
    for entry in entries:
        if skip_hidden and entry.name.startswith(".") and entry.name != ".projhub":
            continue

        if entry.is_dir():
            if entry.name in _SKIP_DIRS:
                continue
            badges = _detect_badges(entry)
            children = scan_dir(entry, depth + 1, max_depth, skip_hidden) if depth < max_depth else None
            result.append({
                "name": entry.name,
                "type": "dir",
                "path": entry.name,
                "badges": badges,
                "hasChildren": True,
                "children": children,
            })
        else:
            result.append({
                "name": entry.name,
                "type": "file",
                "path": entry.name,
                "badges": [],
                "ext": entry.suffix.lstrip(".").lower(),
            })

    result.sort(key=lambda x: (0 if x["type"] == "dir" else 1, x["name"].lower()))
    return result

File: src/lib/filetree.js
import fs from 'node:fs';
import path from 'node:path';

const SKIP_DIRS = new Set([
  'node_modules', 'dist', 'build', '.next', 'out', 'target',
  '__pycache__', '.venv', 'venv', '.tox', 'coverage',
  '.turbo', '.parcel-cache', '.cache',
]);

const BADGE_RULES = [
  { files: ['.git'], badge: 'git', label: 'Git' },
  { files: ['package.json'], badge: 'node', label: 'Node', detect: detectNodeFramework },
  { files: ['pyproject.toml', 'requirements.txt', 'setup.py'], badge: 'python', label: 'Python' },
  { files: ['go.mod'], badge: 'go', label: 'Go' },
  { files: ['Cargo.toml'], badge: 'rust', label: 'Rust' },
  { files: ['pubspec.yaml'], badge: 'flutter', label: 'Flutter' },
  { files: ['build.gradle', 'pom.xml', 'build.gradle.kts'], badge: 'java', label: 'Java' },
  { files: ['composer.json'], badge: 'php', label: 'PHP' },
  { files: ['Dockerfile', 'docker-compose.yml', 'docker-compose.yaml'], badge: 'docker', label: 'Docker' },
  { files: ['terraform'], badge: 'terraform', label: 'Terraform', isDir: true },
];

function detectNodeFramework(dirPath) {
  try {
    const pkg = JSON.parse(fs.readFileSync(path.join(dirPath, 'package.json'), 'utf8'));
    const deps = { ...pkg.dependencies, ...pkg.devDependencies };
    if (deps['react'] || deps['next']) return 'React';
    if (deps['vue'] || deps['nuxt']) return 'Vue';
    if (deps['svelte'] || deps['@sveltejs/kit']) return 'Svelte';
    if (deps['express'] || deps['fastify'] || deps['hono']) return 'Server';
    if (deps['react-native'] || deps['expo']) return 'React Native';
    if (deps['electron']) return 'Electron';
  } catch {}
  return 'Node';
}

function detectBadges(dirPath) {
  const entries = new Set(
    fs.existsSync(dirPath)
      ? fs.readdirSync(dirPath).map(e => e.toLowerCase())
      : []
  );
  const badges = [];

  for (const rule of BADGE_RULES) {
    const hit = rule.files.some(f => {
      if (rule.isDir) return entries.has(f.toLowerCase()) &&
        fs.statSync(path.join(dirPath, f)).isDirectory();
      return entries.has(f.toLowerCase());
    });
    if (hit) {
      const label = rule.detect ? rule.detect(dirPath) : rule.label;
      badges.push({ badge: rule.badge, label });
    }
  }

  // Detect .tf files for terraform (without terraform/ dir)
  if (!badges.find(b => b.badge === 'terraform')) {
    const hasTf = [...entries].some(e => e.endsWith('.tf'));
    if (hasTf) badges.push({ badge: 'terraform', label: 'Terraform' });
  }

  // Detect Xcode
  const hasXcode = [...entries].some(e => e.endsWith('.xcodeproj') || e.endsWith('.xcworkspace'));
  if (hasXcode) badges.push({ badge: 'ios', label: 'iOS' });

  return badges;
}

export function scanDir(absPath, opts = {}) {
  const { depth = 0, maxDepth = 1, skipHidden = false } = opts;

  if (!fs.existsSync(absPath)) return [];

  let entries;
  try {
    entries = fs.readdirSync(absPath, { withFileTypes: true });
  } catch {
    return [];
  }

  const result = [];

  for (const entry of entries) {
    if (skipHidden && entry.name.startsWith('.') && entry.name !== '.holoctl') continue;

    const entryPath = path.join(absPath, entry.name);
    const relPath = entry.name;

    if (entry.isDirectory()) {
      if (SKIP_DIRS.has(entry.name)) continue;

      const badges = detectBadges(entryPath);
      const node = {
        name: entry.name,
        type: 'dir',
        path: relPath,
        badges,
        hasChildren: true,
        children: depth < maxDepth ? scanDir(entryPath, { ...opts, depth: depth + 1 }) : null,
      };
      result.push(node);
    } else {
      result.push({
        name: entry.name,
        type: 'file',
        path: relPath,
        badges: [],
        ext: path.extname(entry.name).slice(1).toLowerCase(),
      });
    }
  }

  // Dirs first, then files, both alphabetically
  result.sort((a, b) => {
    if (a.type !== b.type) return a.type === 'dir' ? -1 : 1;
    return a.name.localeCompare(b.name);
  });

  return result;
}
